test/library/packages/Shlex/Shlex.chpl

```diff
@@ -0,0 +1,385 @@
+            else{
+              if(this.whitespace.find(nextchar)==-1){
+                this._pushback_chars.append(nextchar);
+              }
+              this.state=' ';
+              break;
+            }
+          }
+          else if(this.posix && this.quotes.find(nextchar)!=-1){
+            this.state=nextchar;
+          }
+          else if(this.posix && this.escapse.find(nextchar)!=-1){
+            escapsedstate='a';
+            this.state=nextchar;
+          }
+          else if((this.wordchars.find(nextchar)!=-1) ||
+                    (this.quotes.find(nextchar)!=-1) || this.whitespace_split)
+          {
+            this.token+=nextchar;
+          }
+          else{
+            if(this.punctuation_chars!=''){
+              this._pushback_chars.append(nextchar);
+            }
+            else{
+              this.pushback.insert(0, nextchar);
+            }
+            if this.debug>=2{
+              writeln("shlex: I see punctuation in word state");
+            }
+            this.state=' ';
+            if(this.token!='' || (this.posix && quoted)) then break;
+            else continue;
+          }
+        }
+      }
+      var result = this.token;
+      this.token='';
+      if(this.debug>=2){
+        if result!='' {
+          writeln("shlex: raw token: "+result);
+        }
+        else{
+          writeln("shlex: raw token= EOF");
+        }
+      }
+      return result;
+    }
+  }
+
+  //
+  // This is the public interface to the SHLEX.
+  // It splits a string into a list of tokens in accordance to the rules
+  // specified by the Shlex object.
+  //
+  proc split(s:string, comments:bool = false, posix:bool=true,
+              punctuation_chars: bool = false,
+                custom_punctuation_chars: string = "")
+  {
+    var commenters='';
+    if comments then commenters='#';
+    var lex=new shlex(s, posix=posix, punctuation_chars=(punctuation_chars || custom_punctuation_chars!=''), commenters=commenters, custom_punctuation_chars=custom_punctuation_chars);
+    lex.whitespace_split=!(punctuation_chars || custom_punctuation_chars !='');
+    var lst: list(string);
+    var pres:string = ' ';
+    while(lex.tokindex!=-1 || !lex.pushback.isEmpty()){
+      pres=lex.get_token();
+      if(posix || pres!=''){
+        lst.append(pres);
+      }
+    }
+    return lst;
+  }
+
+  // Regular expression for checking if a string is a token by itself.
+  var _find_unsafe = try! compile("[^\\w@%+=:,./-]");
+
+  //
+  // This is the public interface to the SHLEX.
+  // It returns a shell-escaped version of the string s.
+  // The returned string can be safely used as a token in command-line.
+  //
+  proc quote(s:string){
+    if(s==''){
+      return '""';
+    }
+    if(!s.search(_find_unsafe)){
+      return s;
+    }
+    return "'" + s.replace("'", "'\"'\"'") + "'";
+  }
+
+  //
+  // This is the public interface to the SHLEX.
+  // Concatenates the tokens of the list passed into a string.
```

Comment:
Check out [chpldoc](https://chapel-lang.org/docs/master/tools/chpldoc/chpldoc.html) if you haven't already. We can generate documentation from the module if you use the format:

```
/*

documentation

*/
proc join(...) ... {
   ...
}